### PR TITLE
Re-organized fields, meshes and introduced constraint

### LINKED ISSUE #4: Refactor: structure current high level into packages
Currently, the code is organized into a main `src/hidenn_playground/` package and sub-modules: `quadratures.py, shape_functions.py`. This should be re-organized to anticipate for additional content by creating instead packages `quadratures/` and moving to individual modules the content of current modules, e.g. change `quadratures.py` into:
```bash
- `quadratures/`
  - `quadrature_rule.py` - contains `QuadratureRule` base class
  - `mid_point_1d.py` - contains `MidPoint1D` class
  - `two_points_1d.py` - contains `TwoPoints1D` class
```

### LINKED ISSUE #7: Second stage refactor: improve field interface.
As a second stage of refactor, current `Field` interface should be generalized: better naming of free/imposed dof values. Consider whether boundary condition handling should be part of this class or separate. For now, having it as part of the `Field` class makes it easy to use for Dirichlet boundary condition but for periodic BC it might be more challenging and requires a separate interface.

### PR DESCRIPTION
**Pull request by jdichamp**
This PR addresses the following items:
* Split `Field` into a `Field` and a `TrainableField`. Generalized Dirichlet boundary conditions by having `TrainableField` with a `Constraint` attribute which is responsible for determining the mask of `dofs_free`. This should be straightforward for periodic boundary conditions by defining slaves and master nodes and imposing values slaves = master. (#7)
* `Mesh`is now split into topology and positions to prepare for (#10). In more details it is now:
  * A `Topology` - nodes indices and connectivity
  * An array of nodes' positions
* `Evaluator` is now renamed to `Interpolator` and its implementation is improved (split into helper functions, simplified implementation).
* Re-organized the structure by introducing packages (#4). Note that I did not include any package/module at the top model to enforce user to do at least one level include: `from neurom.shape_functions import LinearSegment`. This helps IMHO to better understand the code structure when writing simulation.
* Added some new unit tests for mesh, fields, constraints, etc. (#5)
* Added first version of workflow which checks we can build library, all tests pass and format is OK (#2).
* Copied back parts of the `README.md` of previous release (#3). Will have be to be improved in the future.

Closes #4, #7

### CODE REVIEW

File: docs/src/conf.py [hunk 1]
```diff
@@ -6,9 +6,9 @@
 # -----------------------------------------------------------------------------
 # Project information
 # -----------------------------------------------------------------------------
-project = "python-template"
-author = "Jules Dichamp"
-copyright = "2026, Jules Dichamp"
+project = "neurom"
+author = "Alexandre Daby, et al. [TODO]"
+copyright = "2026, Alexandre Daby"
```
Comment (jdichamp) [inline docs/src/conf.py:11]: We should add other authors.

**Review by jdichamp — COMMENTED** [review]
Good work overall, minor things only.

File: docs/src/conf.py (hunk 1 above)
Comment (AlexandreDabySeesaram) [inline docs/src/conf.py:11]: Should we put all authors in the copyright as well ?

File: src/neurom/constraints/dirichlet.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        values_imposed (torch.Tensor): The values imposed on those nodes.
+
+    Attributes:
+        nodes (torch.Tensor): The nodes which have a Dirichelet BC.
+        values_imposed (torch.Tensor): The values imposed on those nodes.
+    """
+
+    def __init__(self, nodes, values_imposed):
+        super().__init__()
+        self.register_buffer("nodes", torch.as_tensor(nodes))
```
Comment (AlexandreDabySeesaram) [inline src/neurom/constraints/dirichlet.py:24]: Should we explicitly use a name such as "constrained_nodes" ?

File: src/neurom/meshes/mesh.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+        # List elements to which `x` belongs to.
+        ids = []
+        for x_i in x:
+            for e, conn in enumerate(self.topology.connectivity):
+                x_first = self.nodes_positions.full_values()[conn[0]]
+                x_second = self.nodes_positions.full_values()[conn[1]]
+                if x_i >= x_first and x_i <= x_second:
+                    ids.append(e)
+                    break
```
Comment (AlexandreDabySeesaram) [inline src/neurom/meshes/mesh.py:60]: Works well in 1D, might need external tools for 2D and 3D (vtk, ...)

File: src/neurom/meshes/topology.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
+import torch
+import torch.nn as nn
+
+
+class Topology(nn.Module):
```
Comment (AlexandreDabySeesaram) [inline src/neurom/meshes/topology.py:5]: Why not, would the name "connectivity" work as well ?

File: src/neurom/reference_elements/segment.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import torch
+
+from neurom.reference_elements.reference_element import ReferenceElement
+
+
+class Segment(ReferenceElement):
```
Comment (AlexandreDabySeesaram) [inline src/neurom/reference_elements/segment.py]: Could we name those "bar elements" ?

File: src/neurom/interpolator.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+from neurom.geometry import barycentric_to_reference
+
+
+class Interpolator(nn.Module):
+    """Class encapsulating interpolation of a function on a mesh.
+
+    The class provides methods to interpolate a function at arbitrary positions or at quadrature points.
+    """
+
```
Comment (AlexandreDabySeesaram) [inline src/neurom/interpolator.py:12]: Later I guess this class can be used to evaluate also tensor decompositions for given parameteric configurations

**Review by AlexandreDabySeesaram — COMMENTED** [review]
Very nice, mostly naming convention to be discussed

File: docs/src/conf.py (hunk 1 above)
Comment (jdichamp) [reply docs/src/conf.py:11]: We can, I don't know what standard practices regarding this aspect. Is there anyone else besides @mgenet, you and @KaterinaSkardova?

File: src/neurom/constraints/dirichlet.py (hunk 2 above)
Comment (jdichamp) [reply src/neurom/constraints/dirichlet.py:24]: I intentionnally did not at is already part of `constraints` , so IMHO it would be redundant. The hierarchy already explicitly states that this deals with constraints.

File: src/neurom/meshes/mesh.py (hunk 3 above)
Comment (jdichamp) [reply src/neurom/meshes/mesh.py:60]: To be honest, this was just a hacky (and lazy) way of finding the relevant elements. I have a vectorized version in [this commit](https://github.com/AlexandreDabySeesaram/NeuROM/pull/15/changes/71049d38a745fc58904a95951a02b75994c9f27e) in #15 that will come later which is much better IMHO. Also, we can introduce contact detection algorithms later on, like AABBs, etc. 

If you are ok with it I would suggest to keep as such and I will deal with it in #15 along with cleaning the vectorized implementation.

File: src/neurom/meshes/topology.py (hunk 4 above)
Comment (jdichamp) [reply src/neurom/meshes/topology.py:5]: Topology is meant as the graph representation of the mesh, i.e. : the nodes + the connectivity (which are basically vertex indices). I can rename the current `connectivity` attribute for `vertex_indices` eventually and rename `Topology` into `Connectivity` if you like it better.

File: src/neurom/reference_elements/segment.py (hunk 5 above)
Comment (jdichamp) [reply src/neurom/reference_elements/segment.py]: Since it's in `reference_elements` package, are you ok with just `bar`? It would be imported as `import neurom.reference_elements.bar`.

File: src/neurom/interpolator.py (hunk 6 above)
Comment (jdichamp) [reply src/neurom/interpolator.py:12]: Actually I refactored the interpolation (and just pushed new things to it) in #14. The `Interpolator` (which is renamed in #14 in a `QuadratureAssembly` is really the interpolation (assembly) on quadrature points and will depend on a `QuadratureContext` which is responsible for dealing with the famous $\xi \Rightarrow x \Rightarrow \xi$. But we can see later what is necessary for tensor decompositions like you mentioned. Might need a dedicated class/function for this.

File: docs/src/conf.py (hunk 1 above)
Comment (AlexandreDabySeesaram) [reply docs/src/conf.py:11]: So far there's Martin, Katka and you

File: src/neurom/reference_elements/segment.py (hunk 5 above)
Comment (AlexandreDabySeesaram) [reply src/neurom/reference_elements/segment.py]: Yes !

File: src/neurom/meshes/topology.py (hunk 4 above)
Comment (jdichamp) [reply src/neurom/meshes/topology.py:5]: Opinions @AlexandreDabySeesaram?

File: src/neurom/reference_elements/segment.py (hunk 5 above)
Comment (jdichamp) [reply src/neurom/reference_elements/segment.py]: Shall I rename the `shape_functions.linear_segment` and `shape_functions.quadratic_segment` to `shape_functions.linear_bar` and `shape_functions.quadratic_bar.`?

File: docs/src/conf.py (hunk 1 above)
Comment (jdichamp) [reply docs/src/conf.py:11]: Changed. I added authors by alphabetical order, let me know if it should be done otherwise.
